test/vms.destroy.test.js

```diff
@@ -0,0 +1,532 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2015, Joyent, Inc.
+ */
+
+var assert = require('assert-plus');
+var async = require('async');
+var clone = require('clone');
+var ldapjs = require('ldapjs');
+
+var common = require('../lib/common');
+var testCommon = require('./common');
+var workflow = require('./lib/workflow');
+var vmTest = require('./lib/vm');
+
+var client;
+var MORAY = require('../lib/apis/moray');
+
+var IMAGE = 'fd2cc906-8938-11e3-beab-4359c665ac99';
+var CUSTOMER = testCommon.config.ufdsAdminUuid;
+var NETWORKS = null;
+var SERVER = null;
+
+var VMS_LIST_ENDPOINT = '/vms';
+
+var vmLocation;
+var jobLocation;
+var vmUuid;
+var vmObject;
+var leftoverTestVms = [];
+var leftoverTestVmsDestroyJobUuids = [];
+
+exports.setUp = function (callback) {
+    testCommon.setUp(function (err, _client) {
+        assert.ifError(err);
+        assert.ok(_client, 'restify client');
+        client = _client;
+        callback();
+    });
+};
+
+exports.find_headnode = function (t) {
+    client.cnapi.get('/servers', function (err, req, res, servers) {
+        t.ifError(err);
+        t.equal(res.statusCode, 200);
+        t.ok(servers);
+        t.ok(Array.isArray(servers));
+        for (var i = 0; i < servers.length; i++) {
+            if (servers[i].headnode === true) {
+                SERVER = servers[i];
+                break;
+            }
+        }
+        t.ok(SERVER);
+        t.done();
+    });
+};
+
+exports.napi_networks_ok = function (t) {
+    client.napi.get('/networks', function (err, req, res, networks) {
+        t.ifError(err);
+        t.equal(res.statusCode, 200);
+        t.ok(networks);
+        t.ok(Array.isArray(networks));
+        t.ok(networks.length > 1);
+        NETWORKS = networks;
+        t.done();
+    });
+};
+
+/*
+ * Fist, delete any leftover VMs from a previous tests run that may not have
+ * been cleaned up properly.
+ */
+exports.get_leftover_non_destroyed_test_vms = function (t) {
+    client.get(VMS_LIST_ENDPOINT + '?alias=' + vmTest.TEST_VMS_ALIAS +
+        '&state=destroying',
+        function (err, req, res, body) {
+            t.ifError(err);
+            t.equal(res.statusCode, 200);
+            t.ok(Array.isArray(body));
+            leftoverTestVms = body;
+            t.done();
+        });
+};
+
+exports.remove_leftover_test_vms = function (t) {
+    async.each(leftoverTestVms, function (testVm, nextVm) {
```

Comment:
Is it better to use vasync instead of async here? I think in general we're trying to move away from async and to vasync... But if it's hard to do in this module it's fine to do that separately later.